File: eventhubskafka-flink-eventhubskafka/flink-kafka-consumer/src/main/scala/SimpleRelayStreamingJob.scala
package com.microsoft.samples.flink

import java.time.Instant

import com.microsoft.samples.flink.StreamingJobCommon.createKafkaConsumer
import com.microsoft.samples.flink.data.SampleRecord
import com.microsoft.samples.flink.utils.JsonMapperSchema
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.streaming.api.scala._
import org.slf4j.LoggerFactory

/**
 * A Flink Streaming Job that relays incoming events.
 *
 */
object SimpleRelayStreamingJob {

  private val LOG = LoggerFactory.getLogger(getClass)

  def main(args: Array[String]): Unit = {

    val params = ParameterTool.fromArgs(args)

    val env = StreamingJobCommon.createStreamExecutionEnvironment(params)
    val schema = new JsonMapperSchema(classOf[SampleRecord])
    val kafkaIn = StreamingJobCommon.createKafkaConsumer(params, schema)
    val kafkaOut = StreamingJobCommon.createKafkaProducer(params, schema)

    // Create Flink stream source from Kafka.
    val stream = env.addSource(kafkaIn)

    // Build Flink pipeline.
    stream
      .map(e => {
        e.enqueuedAt = e.createdAt // TODO use actual Kafka enqueued time
        e.processedAt = Instant.now
        e
      })
      .addSink(kafkaOut)

    // execute program
    env.execute("simple relay")
  }


}

File: components/apache-flink/flink-kafka-consumer/src/main/java/com/microsoft/samples/flink/SimpleRelayStreamingJob.java
package com.microsoft.samples.flink;

import com.microsoft.samples.flink.data.SampleRecord;
import com.microsoft.samples.flink.utils.DuplicateByKeyFilter;
import com.microsoft.samples.flink.utils.JsonMapperSchema;
import com.microsoft.samples.flink.utils.SampleRecordTSExtractor;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.sink.SinkFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumerBase;
import org.apache.kafka.clients.consumer.ConsumerRecord;

import java.time.Instant;

import static com.microsoft.samples.flink.StreamingJobCommon.getParams;

public class SimpleRelayStreamingJob {
    private static final int MAX_EVENT_DELAY = 60; // max delay for out of order events

    public static void main(String[] args) throws Exception {
        ParameterTool params = getParams(args);

        StreamExecutionEnvironment env = StreamingJobCommon.createStreamExecutionEnvironment(params);
        JsonMapperSchema<SampleRecord> schema = new JsonMapperSchema<>(SampleRecord.class);
        FlinkKafkaConsumerBase<ConsumerRecord<byte[], SampleRecord>> consumer = StreamingJobCommon.createKafkaConsumer(params, schema);
        SinkFunction<SampleRecord> producer = StreamingJobCommon.createKafkaProducer(params, schema);

        // setup streaming execution environment
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.enableCheckpointing(500);
        env.getConfig().setAutoWatermarkInterval(1000);

        // assign a timestamp extractor
        consumer.assignTimestampsAndWatermarks(new SampleRecordTSExtractor(Time.seconds(MAX_EVENT_DELAY)));

        // Build Flink pipeline.
        buildStreamingJob(env.addSource(consumer), producer);

        // execute program
        env.execute("simple relay");
    }

    static void buildStreamingJob(DataStream<ConsumerRecord<byte[], SampleRecord>> s, SinkFunction<SampleRecord> producer) {
        s.map(r -> {
            SampleRecord e = r.value();
            e.enqueuedAt = Instant.ofEpochMilli(r.timestamp());
            e.processedAt = Instant.now();
            return e;
        })
                .keyBy(e -> e.deviceId)
                .flatMap(new DuplicateByKeyFilter<>(e -> e.eventId, 100))
                .addSink(producer);
    }

}
